### PR TITLE
BEEFY subscription fires multiple times for the same commitment

### PR DESCRIPTION
**Pull request by Wizdave97**
Fixes https://github.com/paritytech/grandpa-bridge-gadget/issues/159

### CODE REVIEW

File: client/beefy/rpc/src/lib.rs [hunk 1]
```diff
@@ -393,4 +410,74 @@ mod tests {
...
+		r.unwrap();
+
+		// Inspect what we received
+		// We should have received only two commitments
+		let recvs = futures::executor::block_on(receiver.take(2).collect::<Vec<_>>());
+		let recv: Notification = serde_json::from_str(&recvs[0]).unwrap();
+		let mut json_map = match recv.params {
+			Params::Map(json_map) => json_map,
+			_ => panic!(),
+		};
```
Comment (seunlanlege) [inline client/beefy/rpc/src/lib.rs:455]: Simply compare to a static string the expected response

**Review by seunlanlege — CHANGES_REQUESTED** [review]

File: client/beefy/rpc/src/lib.rs (hunk 1 above)
Comment (Wizdave97) [reply client/beefy/rpc/src/lib.rs:455]: The response string contains a subscription id that is dynamically generated on each run, so the static string might not work.

Comment (seunlanlege) [reply client/beefy/rpc/src/lib.rs:455]: > The response string contains a subscription id that is dynamically generated on each run, so the static string might not work.

no way to mock subscription id?

File: client/beefy/rpc/src/lib.rs [hunk 2]
```diff
@@ -120,6 +120,7 @@ pub struct BeefyRpcHandler<Block: BlockT> {
 	signed_commitment_stream: BeefySignedCommitmentStream<Block>,
 	beefy_best_block: Arc<RwLock<Option<Block::Hash>>>,
 	manager: SubscriptionManager,
+	beefy_best_block_num: Arc<RwLock<Option<NumberFor<Block>>>>,
```
[the remark this answers is not in the document]
Comment (acatangiu) [reply client/beefy/rpc/src/lib.rs:123]: or even:

```suggestion
	beefy_best_block: Arc<RwLock<Option<(Block::Hash, NumberFor<Block>)>>>,
	manager: SubscriptionManager,
```

**Review by acatangiu — APPROVED** [review]
The code looks good in terms of making sure that RPC subscription doesn't forward duplicate (or old) commitments.

It's not clear why there are duplicate commitments to begin with... should we at least log something when seeing these duplicate commitments so we can further investigate?

**Review by bkchr — CHANGES_REQUESTED** [review]
The issue mentions to find out why commitments are send twice and not to just prevent sending them multiple times from the rpc layer.

**Comment by seunlanlege** [comment]
> The issue mentions to find out why commitments are send twice and not to just prevent sending them multiple times from the rpc layer.

@bkchr 

Because of the way networking works, peers could still be broadcasting votes for a round that has already concluded. The best we can do is dedupe the notifications sent to the rpc layer.

File: client/beefy/rpc/src/lib.rs [hunk 3]
```diff
@@ -365,7 +365,7 @@ mod tests {
 		);
 	}
 
-	fn create_commitment(block_number: u32) -> BeefySignedCommitment<Block> {
+	fn create_commitment(block_number: u64) -> BeefySignedCommitment<Block> {
```
Comment (seunlanlege) [inline client/beefy/rpc/src/lib.rs:368]: block numbers are actually u32

Comment (Wizdave97) [reply client/beefy/rpc/src/lib.rs:368]: The `susbtrate-test-runtime` uses a u64

**Comment by Wizdave97** [comment]
Bump

**Comment by acatangiu** [comment]
Fixed in https://github.com/paritytech/substrate/pull/10882

**Pull request closed without merging**